### DISCUSSION TITLE
Automatic save

### DISCUSSION THREAD

**Discussion by bago** [voidlabs/mosaico#628] [Ideas]
We bundle localstorage saving by default: maybe it should be simply automatic and not triggered by "Save" button.

**Comment by bwl21** [comment]
Save button should alle the user to chose a Name in the localstore, while autosave does not. You could trigger it when you update the undo. And you could also save the undo stack.

**Comment by bago** [comment]
At this time a random name is assigned to each content by "index.html" and then the "editor.html" is launched referencing the generated "name" (you see the key as #hashdata in the url).

Given localstorage is not for "Long term" then we could simply save every action (like google docs). And you are right, we could even save the undo/redo stack (as we already made it serializable).

**Comment by bwl21** [comment]
There are actually two reasons to save in localstore:
1. if there is no other storage backend - In this scenario you want to save more than one template. Save is usually done by user interaction.
2. to prevent data loss in case of a browser crash or accidential closeing. In this scneario you want to save your current work. Save should be done automatically.

I have written a music app (http://zupfnoter.weichel21.de) where I followed exactly this approach. It has saved me many times.

I also have a save to dopbox there, but I did not remove the localstore.

**Comment by bago** [comment]
Wow, the music app is really cool!

I also thought that, given we have a backend for the image uploads, we should probably "save" by simply uploading the model+metadata json to the same area where images are uploaded.
No need for more backend "features" and we could leave the localstorage only for "autosave".
When you start the editor it should load the remote "metadata+content" and compare it with the one in the localstorage: if the one in localstorage is newer it should ask if you want to load "unsaved changes in the browser".
